### PR TITLE
refactoring

### PR DESCRIPTION
**Pull request by BlueHotDog**
What do you think? mainly some cleanup, main thing are the transactions.

### CODE REVIEW

File: server/lib/buff_server/packages.ex [hunk 1]
```diff
@@ -29,90 +30,98 @@ defmodule BuffServer.Packages do
...
-
+  Creates the package and uploads to S3, this is run inside a transaction so if anything fails it'll rollback.
   """
-  def create_package!(%{artifact_binary: artifact_binary} = attrs) when is_binary(artifact_binary) do
-    changeset = %Package{}
-    |> Package.changeset(attrs, [:s3_bucket_name, :s3_bucket_path])
-
-    bucket_path = "/#{attrs.name}/artifact"
-    %{status_code: 200} = ExAws.S3.put_object(@s3_bucket_name, bucket_path, artifact_binary) |> ExAws.request!
+  def create_package(%{artifact_binary: artifact_binary} = attrs)
```
Comment (BlueHotDog) [inline server/lib/buff_server/packages.ex]: Its usually more convenient to have method not throw

File: server/lib/buff_server/packages/package.ex [hunk 2]
```diff
@@ -22,41 +30,22 @@ defmodule BuffServer.Packages.Package do
...
-          %URI{scheme: nil} ->
-            {:err, "is missing a scheme (e.g. https)"}
-
-          %URI{host: nil} ->
-            {:err, "is missing a host"}
-
-          %URI{host: host} ->
-            validate_host.(host)
-        end
+    validate_change(changeset, field, fn _, url ->
```
Comment (BlueHotDog) [inline server/lib/buff_server/packages/package.ex]: discovered this, looks more robust..

File: server/lib/buff_server/packages.ex (hunk 1 above)
Comment (itayadler) [reply server/lib/buff_server/packages.ex]: MatchError occurs at compile time?

File: server/lib/mix/tasks/minio_setup.ex [hunk 3]
```diff
@@ -1,38 +1,21 @@
...
   def run(_) do
     region = Application.get_env(:ex_aws, :s3)[:region]
     s3_bucket_name = Application.get_env(:buff_server, :s3_bucket_name)
 
     Mix.Task.run("app.start")
-    Mix.shell.info("Creating bucket #{s3_bucket_name}...")
+    Mix.shell().info("Creating bucket #{s3_bucket_name}...")
 
-    case ExAws.S3.put_bucket(s3_bucket_name, region) |> ExAws.request do
+    case s3_bucket_name |> ExAws.S3.put_bucket(region) |> ExAws.request(http_client: ExAws.Request.Hackney) do
```
Comment (itayadler) [inline server/lib/mix/tasks/minio_setup.ex]: nice! that way it won't throw the error, 👍

File: server/test/buff_server/packages_test.exs [hunk 4]
```diff
@@ -1,97 +1,105 @@
...
-        description: "some description",
-        homepage: "some homepage",
-        keywords: [],
-        name: "some-name",
-        repository_url: "http://aaa.com",
-        artifact_binary: artifact_fixture(@dummy_artifact)
-      }
-    end
+    @s3_bucket_name Application.get_env(:buff_server, :s3_bucket_name)
+    @dummy_artifact File.read("test/buff_server/fixtures/dummy_artifact.gz") |> elem(1)
```
Comment (itayadler) [inline server/test/buff_server/packages_test.exs]: elem(1)? that'll take the 2nd element from a Map?

File: server/test/buff_server/packages_test.exs [hunk 5]
```diff
@@ -1,97 +1,105 @@
...
       name: nil,
-      repository_url: nil
+      repository_url: nil,
+      artifact_binary: @dummy_artifact
     }
 
     def package_fixture(attrs) do
       ExAws.Request.HttpMock
-      |> expect(:request, fn _method, _url, _body, _headers, _opts -> {:ok, %{status_code: 200}} end)
+      |> stub(:request, fn _method, _url, _body, _headers, _opts ->
```
Comment (itayadler) [inline server/test/buff_server/packages_test.exs]: why stub and not expect?

File: server/test/buff_server/packages_test.exs [hunk 6]
```diff
@@ -1,97 +1,105 @@
...
-      assert package.repository_url == attrs().repository_url
+      {:ok, package} = package_fixture(@default_attrs)
+      assert package.name == @default_attrs.name
+      assert package.description == @default_attrs.description
+      assert package.homepage == @default_attrs.homepage
+      assert package.keywords == @default_attrs.keywords
+      assert package.repository_url == @default_attrs.repository_url
       assert package.s3_bucket_name == @s3_bucket_name
-      assert package.s3_bucket_path == "/#{attrs().name}/artifact"
+      assert package.s3_bucket_path == Packages.s3_bucket_path(@default_attrs)
```
Comment (itayadler) [inline server/test/buff_server/packages_test.exs]: 👍

File: server/test/buff_server/packages_test.exs (hunk 5 above)
Comment (itayadler) [reply server/test/buff_server/packages_test.exs]: ah ok so i'll set the expectations explicitly in the test, nice

File: server/test/mix/tasks/minio_setup_test.exs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+defmodule Mix.Tasks.Minio.SetupTest do
+  use BuffServer.MixCase, async: true
+
+  describe "run/1" do
+    test "runs successfully" do
```
Comment (itayadler) [inline server/test/mix/tasks/minio_setup_test.exs]: 👍

File: server/lib/buff_server/packages.ex (hunk 1 above)
Comment (itayadler) [reply server/lib/buff_server/packages.ex]: hmm no that doesn't make any sense, however I don't see why the bang ! convention makes sense if any function case throw a MatchError.

**Comment by itayadler** [comment]
nm merged it to my branch

File: server/test/buff_server/packages_test.exs (hunk 4 above)
Comment (BlueHotDog) [reply server/test/buff_server/packages_test.exs]: i think it takes the first.. lol

File: server/lib/buff_server/packages.ex (hunk 1 above)
Comment (BlueHotDog) [reply server/lib/buff_server/packages.ex]: MatchError happens in run-time. 
yeah, you're kinda right. its super hard to really enforce this in a dynamic lang..
But the bang notation means the function itself shouldnt throw by default..(it still might, i guess :( )

**Pull request closed without merging**